File: 747-min-cost-climbing-stairs/solution.py
class Solution(object):
    def minCostClimbingStairs(self, cost):
        """
        :type cost: List[int]
        :rtype: int
        """
        self.memo = {}

        def dp(i):
            if i <= 1: return 0
            if i in self.memo: return self.memo[i]

            one_step = cost[i-1] + dp(i-1)
            two_step = cost[i-2] + dp(i-2)
            self.memo[i] = min(one_step, two_step)
            return self.memo[i]
        
        return dp(len(cost))

File: 747-min-cost-climbing-stairs/solution.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        return topDownDP(cost, cost.size());
    }
    
    int topDownDP(vector<int>& c, int idx) {
        if ( 1 == dp.count(idx)) return dp[idx];
        if ( idx <= 1 ) return dp[idx] = c[idx];
        return dp[idx] = (idx == c.size() ? 0: c[idx]) + min(topDownDP(c, idx-1), topDownDP(c, idx-2));
    }
    
private:
    unordered_map<int, int> dp;
};
